module-audio/board/linux/PulseAudioWrapper.cpp

```diff
@@ -0,0 +1,159 @@
+        mainloop_api = pa_mainloop_get_api(mainloop);
+        context      = pa_context_new(mainloop_api, "PureOSContext");
+        pa_context_set_state_callback(
+            context,
+            [](pa_context *, void *arg) {
+                auto *inst = static_cast<PulseAudioWrapper *>(arg);
+                inst->context_state_callback();
+            },
+            this);
+        pa_context_connect(context, nullptr, {}, nullptr);
+
+        int ret = 1;
+        if (pa_mainloop_run(mainloop, &ret) < 0) {
+            fprintf(stderr, "pa_mainloop_run() failed.\n");
+            std::abort();
+        }
+
+        return nullptr;
+    }
+
+    void PulseAudioWrapper::context_state_callback()
+    {
+        assert(context);
+
+        switch (pa_context_get_state(context)) {
+        case PA_CONTEXT_CONNECTING:
+        case PA_CONTEXT_AUTHORIZING:
+        case PA_CONTEXT_SETTING_NAME:
+            break;
+
+        case PA_CONTEXT_READY: {
+            int r;
+            pa_sample_spec sample_spec;
+            sample_spec.channels = audio_format.getChannels();
+            sample_spec.rate     = audio_format.getSampleRate();
+            sample_spec.format   = PA_SAMPLE_S16LE;
+
+            if ((stream = pa_stream_new(context, "PureOSStream", &sample_spec, nullptr)) == nullptr) {
+                fprintf(stderr, "pa_stream_new() failed: %s\n", pa_strerror(pa_context_errno(context)));
+                return;
+            }
+
+            pa_stream_set_write_callback(
+                stream,
+                [](pa_stream *, size_t length, void *arg) {
+                    auto *inst = static_cast<PulseAudioWrapper *>(arg);
+                    inst->stream_write_cb(length);
+                },
+                this);
+
+            if ((r = pa_stream_connect_playback(stream, nullptr, nullptr, {}, nullptr, nullptr)) < 0) {
+                fprintf(stderr, "pa_stream_connect_playback() failed: %s\n", pa_strerror(pa_context_errno(context)));
+                return;
+            }
+
+            break;
+        }
+
+        case PA_CONTEXT_TERMINATED:
+            quit(0);
+            fprintf(stderr, "PulseAudio connection terminated.\n");
+            break;
+
+        case PA_CONTEXT_FAILED:
+        default:
+            fprintf(stderr, "Connection failure: %s\n", pa_strerror(pa_context_errno(context)));
+            quit(1);
+        }
+    }
+
+    void PulseAudioWrapper::stream_write_cb(size_t length)
+    {
+        write_cb(length);
+    }
+
+    PulseAudioWrapper::~PulseAudioWrapper()
+    {
+        quit();
+        pthread_join(tid, nullptr);
+
+        if (stream != nullptr) {
+            pa_stream_unref(stream);
+        }
+
+        if (context != nullptr) {
+            pa_context_unref(context);
+            context = nullptr;
+        }
+
+        if (mainloop != nullptr) {
+            pa_mainloop_free(mainloop);
+            mainloop     = nullptr;
+            mainloop_api = nullptr;
```

Comment:
is it possible to mainloop_api to not be freed here? or does pa_mainloop_free does that as well?

Reply:
From the documentation, it seems pa_mainloop_get_api`does not allocate anything. It just returns a pointer to some internal abstraction layer. That's how it is done in all examples I've found.